20201022

litehed opened the pull request:
Hi jackson ;)

None of the controls worked and I don't know whether it was a hardware or code error because it wouldn't even move forward if I wanted it to.

litehed committed:
We bringing command based

litehed committed:
Delete TeleOpONLY.java

litehed committed:
CommandBased Test

Begun working on code for command base with the setup of a Drive Subsytem and a command. Still NOT finished.

litehed committed:
Beep Boop MotorEx to Motor

- Changed all instances of MotorEx to Motor
- Also added an outline for the Robot class

litehed committed:
CrimsonRobot was trash so now its gone

litehed committed:
Basis for more commands

+ Added shooter and commands to start and stop
+ Fixed some bugs
+Notice me Jackson-senpai

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOpMain.java:
```diff
@@ -0,0 +1,61 @@
+
+        m_driverOp = new GamepadEx(gamepad1);
+
+        driveCommand = new Com_Drive(mecDrive, ()->m_driverOp.getLeftX(), ()->m_driverOp.getLeftY(), ()->m_driverOp.getRightX());
```
```java
driveCommand = new Com_Drive(mecDrive, m_driverOp::getLeftX, m_driverOp::getLeftY, m_driverOp::getRightX);
```

JIceberg replied:
... bro ...

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/DriveSystem.java:
```diff
@@ -0,0 +1,29 @@
+        drive = new MecanumDrive(fL, fR, bL, bR);
+    }
+
+    public DriveSystem(HardwareMap hMap, String fLName, String fRName, String bLName, String bRName){
+        this(new Motor(hMap, fLName), new Motor(hMap, fRName), new Motor(hMap, bLName), new Motor(hMap, bRName));
+    }
```
Why is this here if you're not using it?

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/Shooter.java:
```diff
@@ -0,0 +1,22 @@
+import com.arcrobotics.ftclib.hardware.motors.MotorEx;
+import com.qualcomm.robotcore.hardware.HardwareMap;
+
+public class Shooter extends SubsystemBase {
+    private Motor kuunav;
+
+    public Shooter(HardwareMap hmap, String name){
+        kuunav = new Motor(hmap, name);
+    }
+
+    public void shot(){
+        kuunav.set(1);
+    }
+
+    public void stoop(){
+        kuunav.set(0);
+    }
+}
```
Your shooter should not create a new Motor reference because Java handles these poorly. Instead we can utilize what I like to call "pseudo-pointers," since under-the-hood these values function as pointers but don't really act like them post-abstraction.
Disregarding that discussion, it would be better practice if you passed a Motor object into the constructor and created that motor in your OpMode.

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOpMain.java:
```diff
@@ -0,0 +1,61 @@
+@TeleOp(name = "CommandBaseTest")
+public class TeleOpMain extends CommandOpMode {
+
+    private Motor fL, bL, fR, bR;
```
Add a shooter motor to pass into the new constructor, reference one of the other comments.

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/commands/Com_NoShoot.java:
```diff
@@ -0,0 +1,23 @@
+
+    private final Shooter shooter;
+
+    public Com_NoShoot(Shooter subby){
+        shooter = subby;
+        addRequirements(shooter);
+    }
+    @Override
+    public void initialize(){
+        shooter.stoop();
+    }
+    @Override
+    public boolean isFinished(){
+        return true;
+    }
```
You need an `execute()` method. Remove the `isFinished()` and replace `initialize()` with `execute()` (the name of the method should be changed).

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/commands/Com_Shoot.java:
```diff
@@ -0,0 +1,23 @@
+        shooter = subby;
+        addRequirements(shooter);
+    }
+    @Override
+    public void initialize(){
+        shooter.shot();
+    }
+    @Override
+    public boolean isFinished(){
+        return true;
+    }
```
You need an `execute()` command. Remove `isFinished()` and replace `initialize()` with `execute()` (the name of the method should be changed).

JIceberg reviewed (changes requested):
Please review the comments I made.

JIceberg commented:
@liteheded pls bro implement the changes and push

litehed committed:
pr changes and update

I updated FTCLib and made changes requested by Jackson This would be really awkward if I missed something

JIceberg reviewed (changes requested):
Change the lambdas to method reference, and then it's good.

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/Shooter.java:
```diff
@@ -0,0 +1,22 @@
+
+    public Shooter(HardwareMap hmap, String name){
+        kuunav = new Motor(hmap, name);
+    }
```
Pass a motor object, not hardware map and string name. There is no point, just pass in the motor.

JIceberg reviewed (changes requested):
Please also look at the suggestion for passing in a motor to the constructor.

litehed committed:
hey uwu

Quick Fix owo

JIceberg reviewed (approved):
Nice, hopefully this works when it's tested tomorrow
